packages/modules/vehicles/smarteq/api.py

```diff
@@ -61,160 +116,266 @@ def set_vin(self, vin: str):
-            # self.log.debug("response3.text = " + str(response3.text))
-            result_json = json.loads(str(bs4.BeautifulSoup(response3.text, 'html.parser')))
-            # self.log.debug("Step3 result_json:\n" + json.dumps(result_json))
-            token = result_json['token']
-            # self.log.debug("login - step3 - get token = " + token)
+            response = self.session.post(url, headers=headers, data=data)
+            self.log.debug("login: status_code = " + str(response.status_code))
+            if response.status_code >= 400:
+                self.log.error("login: failed, status_code = " + str(response.status_code) +
+                               ", check username/password")
+                token = ""
+            else:
+                result_json = json.loads(str(bs4.BeautifulSoup(response.text, 'html.parser')))
+                self.log.debug("login: result_json:\n" + json.dumps(result_json))
+                token = result_json['token']
+                self.log.debug("login: token = " + token)
         except Exception as e:
-            self.log.exception('Step3 Exception: ' + str(e))
+            self.log.error('login:  Exception: ' + str(e))
+            token = ""
         return token
 
     # get code
-    def get_code(self, token: str) -> str:
-        url4 = BASE_URL + '/' + self.resume
-        headers4 = {
+    def get_code(self) -> str:
+        token = self.login()
+        if token == "":
+            self.log.error("login: Login failed - check username/password")
+            return ""
+        url = BASE_URL + '/' + self.resume
+        headers = {
             "Content-Type": "application/x-www-form-urlencoded",
             "Accept": "application/json, text/plain, */*",
             "User-Agent": "Mozilla/5.0 (iPhone; CPU iPhone OS 12_5_1 like Mac OS X)\
             AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148",
             "Referer": LOGIN_URL,
             "Accept-Language": ACCEPT_LANGUAGE,
         }
-        d = {}
-        d['token'] = token
-        data4 = json.dumps(d)
+        data = json.dumps({'token': token})
 
         try:
-            response4 = self.session.post(url4, headers=headers4, data=data4)
-            code = response4.url.split('?')[1].split('=')[1]
-            # self.log.debug("login - step4 - get code = " + code)
+            response = self.session.post(url, headers=headers, data=data)
+            code = response.url.split('?')[1].split('=')[1]
+            self.log.debug("get_code: code=" + code)
         except Exception as e:
-            self.log.exception("Step4 Exception: " + str(e))
+            self.log.error("get_code: Exception: " + str(e))
         return code
 
     # get Tokens
-    def get_Tokens(self, code: str) -> dict:
-        url5 = TOKEN_URL
-        headers5 = {
+    def get_tokens(self) -> dict:
+        self.method += " 3-full (re)connect"
+        code = self.get_code()
```

Comment:
get_tokens ruft get_code ruft login ruft get_resume auf. 
get_code, login, get_resume werden an keiner anderen Stelle aufgerufen. Daher ist folgendes besser lesbar:
```suggestion
        code = self.get_code()
        login_state = self.login(code)
        resume = self.get_resume(login_state)
```
Prüfungen, wie zB in 190-192, können dann auch hier erfolgen. Im Sinne des Single-Responsibility-Patterns sollte die get_code-Methode auch nicht prüfen, ob der Login erfolgreich war, sondern lediglich den gesuchten Code zurückliefern.

Reply:
Es sollten alle Änderungen gemacht sein - bitte noch mal reviewen